### DISCUSSION TITLE
QCefViewTest 的拖拽区域

### DISCUSSION THREAD

**Discussion by L-Super** [CefView/QCefView#275] [Q&A] [answered]
QCefViewTest里，有一块拖拽区域
![image](https://github.com/CefView/QCefView/assets/40905056/2550a4c7-cb3b-4b85-bfa4-e8412e19fee7)
并且代码里也有相关的函数
```
CefViewWidget::onDraggableRegionChanged(const QRegion& draggableRegion, const QRegion& nonDraggableRegion)
```
我在尝试拖拽功能时，无法将文件拖拽到窗口里。所以，想问一下拖拽功能是否支持？如果不支持，可以怎样去实现呢

**Comment by tishion** [comment]
这是拖拽窗口的意思

**Comment by L-Super** [comment]
> 这是拖拽窗口的意思

我尝试了一下按住这区域也不能拖动窗口，不太理解拖拽窗口的意思

**Reply by tishion** [reply]
web内定义一个区域，可以移动当前窗口

**Comment by L-Super** [comment]
那么，拖拽功能有实现吗？如果没有，方便提供一些实现的思路吗
场景：拖拽文件到网页的上传操作

**Comment by L-Super** [comment]
那么，拖拽功能有实现吗？如果没有，方便提供一些实现的思路吗
场景：拖拽文件到网页的上传操作

**Comment by tishion** [comment, marked as answer]
你想实现的功能叫做Drag & Drop （DND），
DND在每个平台/框架都有一套自己的实现
例如你通过native的Windows窗体实现DND，你需要参考：
https://learn.microsoft.com/en-us/windows/win32/shell/dragdrop

如果你想通过Qt框架来实现Qt窗口的DND，你需要参考：
https://doc.qt.io/qt-6/dnd.html


CEF在NCW模式下，已经对接了系统的DND接口，并且会通过CefDragHandler::OnDragEnter方法来告诉你当前用户有DRAG操作。
在OSR模式下，CEF没有提供任何接口，所以完全需要你自己去是实现（通过Qt感知用户的DND事件，然后传给CEF)。

**Comment by robert2010** [comment]
不使用离屏模式，使用窗口模式，页面中的可拖动区域，按住无法拖动，麻烦能提供一些思路？离屏模式是正常的

**Reply by VisualTravel** [reply]
创建个元素，在点击元素时向C++发送信号，C++接收到后使用WM_LBUTTONDOWN和WM_LBUTTONUP监测左键是否按下/抬起，用QCursor获取鼠标在屏幕上的坐标，然后用屏幕上的坐标减去geometry().topLeft()计算窗口内坐标，用屏幕上的坐标减去窗口内坐标进行move

**Comment by L-Super** [comment]
> 不使用离屏模式，使用窗口模式，页面中的可拖动区域，按住无法拖动，麻烦能提供一些思路？离屏模式是正常的
> 

也许可以按照Qt的逻辑去做，判断按下是否属于该区域，然后进行move操作
